compiler/rustc_ast_lowering/src/index.rs

```diff
@@ -54,7 +54,18 @@ pub(super) fn index_hir<'hir>(
         OwnerNode::ForeignItem(item) => collector.visit_foreign_item(item),
     };
 
-    (collector.nodes, collector.parenting)
+    let err = || {
+        let span = item.span();
+        tcx.dcx().span_delayed_bug(*span, "ID not encountered when visiting item HIR");
+        ParentedNode { parent: ItemLocalId::new(0), node: Node::Err(span) }
+    };
+    let nodes = collector
+        .nodes
+        .into_iter()
+        .map(|parented_node| parented_node.unwrap_or_else(err))
+        .collect();
```

Comment:
Should we pre-fill `nodes` with `(0, Node::Err)`, and have the indexer overwrite it, avoiding the re-allocation?